### PR TITLE
Madison/datastructures

### PR DESCRIPTION
**Pull request by madisonmay**

### CODE REVIEW

**Comment by madisonmay** [comment]
This isn't ready for code review and is not passing tests, but wanted to put this up in advance to use Github's diff viewer.

File: finetune/encoding.py [hunk 1]
```diff
@@ -18,8 +18,21 @@
...
+EncoderOutput = namedtuple("EncodedOutput", [
+    "token_ids", # list of list of subtoken ids (ints)
+    "tokens",    # list of list of subtokens (strs)
+    "labels",    # list of list of labels 
+    "char_locs"  # list of list of character locations (ints)
+])
+EncoderOutput.__new__.__defaults__ = (None,) * len(EncoderOutput.fields)
+ArrayEncodedOutput = namedtuple("ArrayEncodedOutput", [
+    "token_ids", # int array shape (batch, seq_length)
+    "mask",      # int array shape (batch, seq_length)
```
Comment (madisonmay) [inline finetune/encoding.py]: Reorder to be consistent

File: finetune/utils.py [hunk 2]
```diff
@@ -13,6 +14,18 @@
 from finetune import config
 
 
+def datatype(typename, field_names):
```
Comment (madisonmay) [inline finetune/utils.py]: need to credit stackoverflow here

File: finetune/base.py [hunk 3]
```diff
@@ -293,9 +291,23 @@ def _predict_proba(self, *Xs, max_length=None):
                 )
                 probas = output.get(self.predict_proba_op)
                 classes = self.label_encoder.target_dim
-                predictions.extend([
-                    dict(zip(classes, proba)) for proba in probas
-                ])
+                if len(probas.shape) == 2:
```
Comment (benleetownsend) [inline finetune/base.py]: Can we not use the inferred type that was set at train time?

File: finetune/base.py [hunk 4]
```diff
@@ -308,7 +320,11 @@ def _featurize(self, *Xs, max_length=None):
             warnings.filterwarnings("ignore")
             max_length = max_length or self.config.max_length
             for xmb, mmb in self._infer_prep(*Xs, max_length=max_length):
-                feature_batch = self.sess.run(self.features, {self.X: xmb, self.M: mmb})
+                feature_batch = self.sess.run(self.features, {
```
Comment (benleetownsend) [inline finetune/base.py]: Interesting, was this broken or has something changed with the logic to add dropout ops?

File: finetune/utils.py (hunk 2 above)
Comment (benleetownsend) [reply finetune/utils.py]: This will ignore misspelt or extra key names. Also, I dont understand, named tuples accept kwargs??

Comment (benleetownsend) [reply finetune/utils.py]: Im sure ive just missed something here, but I dont know what advantage this gives over named tuple.

File: finetune/base.py (hunk 3 above)
Comment (benleetownsend) [reply finetune/base.py]: Or do this in the subclass relating to sequence labeling.
